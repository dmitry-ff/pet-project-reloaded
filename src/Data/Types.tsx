export type TUserData = {
  data: {
    labels: string[];
    datasets: {
      label: string;
      data: number[];
      backgroundColor: CanvasPattern[];
      borderColor: string[];
      borderWidth: number;
      hoverOffset: number;
    }[];
  }
};
export type TLineData = {
  data: {
    labels: string[];
    datasets: {
      label: string;
      data: number[];
      borderColor: string;
      backgroundColor: string;
    }[];
  }
};

export type TLoading = {
  loading: boolean;

};
export type TContext = [
  context: boolean,
  setContext: React.Dispatch<React.SetStateAction<boolean>>,
];
export type TData = {
  date: string,
  day: {
    maxtemp_c: number,
    mintemp_c: number,
    avgtemp_c: number,
    maxwind_kph: number,
    daily_chance_of_rain: number,
    daily_chance_of_snow: number,
    condition: {
      text: string,
      icon: string
    }
  }
}

export type TResponse = {
  forecast: {
    forecastday:
    {
      date: string,
      day: {
        maxtemp_c: number,
        mintemp_c: number,
        avgtemp_c: number,
        maxwind_kph: number,
        daily_chance_of_rain: number,
        daily_chance_of_snow: number,
        condition: {
          text: string,
          icon: string
        }
      },
    }[]
  }
}