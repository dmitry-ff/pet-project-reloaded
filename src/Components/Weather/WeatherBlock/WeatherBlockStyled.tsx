import { css } from '@emotion/react';
import styled from '@emotion/styled';
import rain from '../../../Images/7795646_weather_cloud_rain_icon.svg';
import snow from '../../../Images/8664916_snowflake_icon.svg';
export const WeatherData = styled.div`
    margin-bottom: 35px;
    border-radius: 10px;
    padding:20px;
    box-shadow: 0 0 10px rgba(219, 219, 219, 0.596);
    display: flex;
    margin:auto;
    flex-direction: column;
    align-content:center;
    height: 300px;
    width:max-content;
    min-width:600px;
    justify-content: space-between;
    >span{
        font-size:1.5em;
    }
  
`

export const WeatherToday = styled.div`
    display: flex;
    align-items:flex-end ;
    justify-content: space-between;
    div {
        margin-right:20px;
        display:flex;
    }
`
export const FirstColumn = styled.div`
    display: flex;
    align-items: flex-end;
    & > span{
        font-size: 3em;
    }

`

export const SecondColumn = styled.div`
    display: flex;
    flex-direction: column;
    div{
        display: flex;
        justify-content: space-between;
        width: 150px;
    }
`

const baseStyle = css`
    position: absolute;
    content:'';
    right:-20px;
    
    display: inline-block;
    width: 17px;
    height: 17px;
    background-repeat: no-repeat;
    padding: 0;
`;

export const FourthColumn = styled.div`
    display: flex;
    flex-direction: column;
    align-items:flex-end;
    justify-content:space-between;

    & :first-of-type {
        position: relative;
        &:after{
            vertical-align: top;
            background-image:url(${rain});
            ${baseStyle};
        }
    }
    & :last-of-type {
        position: relative;
        &:after{
            vertical-align: center;
            background-image: url(${snow});
            ${baseStyle};
        }
    }
`


export const WeatherWeek = styled.div`
    display: flex;
    justify-content: space-evenly;
    > div{
        width: 100px;
        border-radius: 15px;
        padding: 15px 20px;
        box-shadow: 5px 5px 10px rgba(211, 211, 211, 0.712);
        background-color: rgba(225, 241, 245, 0.705);
    }
`

export const WeatherCard = styled.div`
        display: flex;
        > div{
            justify-content: space-around;
            display: flex;
            flex-direction: column;
            & :first-of-type{
               font-size:1.5em;
            }
            & :last-of-type{
                font-size:0.8em;   
                color: rgb(102, 102, 102);
            }
        }

`