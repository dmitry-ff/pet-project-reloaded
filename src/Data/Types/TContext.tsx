export type TContext = {
  context: boolean;
  setContext: React.Dispatch<React.SetStateAction<boolean>>;
};
