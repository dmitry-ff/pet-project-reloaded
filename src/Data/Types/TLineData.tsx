export type TLineData = {
  data: {
    labels: string[];
    datasets: {
      label: string;
      data: number[];
      borderColor: string;
      backgroundColor: string;
    }[];
  };
};
