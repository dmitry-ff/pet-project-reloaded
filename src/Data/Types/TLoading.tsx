export type TLoading = {
  loading: boolean;

};
