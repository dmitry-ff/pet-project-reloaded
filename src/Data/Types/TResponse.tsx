export type TResponse = {
  forecast: {
    forecastday: {
      date: string;
      day: {
        maxtemp_c: number;
        mintemp_c: number;
        avgtemp_c: number;
        maxwind_kph: number;
        daily_chance_of_rain: number;
        daily_chance_of_snow: number;
        condition: {
          text: string;
          icon: string;
        };
      };
    }[];
  };
};
