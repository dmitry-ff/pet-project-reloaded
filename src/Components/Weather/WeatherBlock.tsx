import _, { isEmpty } from 'lodash';
import React from 'react';
import { WEEK_S } from "../../Data/Consts/WEEK_S";
import { WeatherData, WeatherToday, FirstColumn, SecondColumn, FourthColumn, WeatherWeek, WeatherCard } from '../../Style/Weather'
import { TLoading } from "../../Data/Types/TLoading";
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
type THead = {
  nameHead: string;
}
type TData = {
  data: {
    date: string,
    day: {
      maxtemp_c: number,
      mintemp_c: number,
      avgtemp_c: number,
      maxwind_kph: number,
      daily_chance_of_rain: number,
      daily_chance_of_snow: number,
      condition: {
        text: string,
        icon: string
      }
    }
  }[]
}

const WeatherBlock: React.FC<TLoading & THead & TData> = ({ loading, nameHead, data }): EmotionJSX.Element | null => {
  console.log(loading, 'LOADING')
  console.log(data, 'DATA');

  if (!loading && isEmpty(data)) return null

  return (
    <WeatherData>
      <span >
        Weather in <b>{nameHead}</b>
      </span>
      <WeatherToday>
        <FirstColumn>
          <span>{_.head(data)!.day.avgtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.avgtemp_c) : _.floor(_.head(data)!.day.avgtemp_c)}°</span>
          <img src={_.head(data)!.day.condition.icon} alt={_.head(data)!.day.condition.text} />
        </FirstColumn>
        <SecondColumn>
          <span>{_.head(data)!.day.condition.text}</span>
          <div>
            <span>max <b>{_.head(data)!.day.maxtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.maxtemp_c) : _.floor(_.head(data)!.day.maxtemp_c)}</b></span>
            <span>min <b>{_.head(data)!.day.mintemp_c > 0 ? '+' + _.floor(_.head(data)!.day.mintemp_c) : _.floor(_.head(data)!.day.mintemp_c)}</b></span>
          </div>
        </SecondColumn>
        <div>
          <span>Wind speed  <b>{(_.head(data)!.day.maxwind_kph / 3.6).toFixed(1)} m/s</b></span>
        </div>
        <FourthColumn>
          <span>{_.head(data)!.day.daily_chance_of_rain} %</span>
          <span>{_.head(data)!.day.daily_chance_of_snow} %</span>
        </FourthColumn>
      </WeatherToday>

      <WeatherWeek>
        {!loading && <>
          {
            _.map(data, (item, index) =>
              <div key={index}>
                <div>
                  {index === 0 ?
                    'Today'
                    : `${WEEK_S[new Date(item.date).getDay()]}, ${new Date(item.date).getDate()} `}
                </div>
                <WeatherCard>
                  <div>
                    <span>{item.day.maxtemp_c > 0 ? '+' + _.floor(item.day.maxtemp_c) : _.floor(item.day.maxtemp_c)}°</span>
                    <span>{item.day.mintemp_c > 0 ? '+' + _.floor(item.day.mintemp_c) : _.floor(item.day.mintemp_c)}°</span>
                  </div>
                  <img src={item.day.condition.icon} alt={item.day.condition.text} />
                </WeatherCard>
              </div>)
          }
        </>
        }

      </WeatherWeek>
    </WeatherData>
  )

}

export default WeatherBlock;
// export { }