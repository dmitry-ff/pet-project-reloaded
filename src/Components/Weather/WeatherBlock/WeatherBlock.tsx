import _, { isEmpty } from 'lodash';
import { WEEK_S } from "../data/WEEK_S";
import * as Styled from './WeatherBlock.styled'
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
import { TData } from '../types/TData'



const WeatherBlock = (props: { loading: boolean, nameHead: string, data: TData[] }): EmotionJSX.Element | null => {
  const { loading, nameHead, data } = props;
  if (!loading && isEmpty(data)) return null;

  return (
    <Styled.WeatherData>
      <span >
        Weather in <b>{nameHead}</b>
      </span>
      <Styled.WeatherToday>
        <Styled.FirstColumn>
          <span>{_.head(data)!.day.avgtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.avgtemp_c) : _.floor(_.head(data)!.day.avgtemp_c)}°</span>
          <img src={_.head(data)!.day.condition.icon} alt={_.head(data)!.day.condition.text} />
        </Styled.FirstColumn>
        <Styled.SecondColumn>
          <span>{_.head(data)!.day.condition.text}</span>
          <div>
            <span>max <b>{_.head(data)!.day.maxtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.maxtemp_c) : _.floor(_.head(data)!.day.maxtemp_c)}</b></span>
            <span>min <b>{_.head(data)!.day.mintemp_c > 0 ? '+' + _.floor(_.head(data)!.day.mintemp_c) : _.floor(_.head(data)!.day.mintemp_c)}</b></span>
          </div>
        </Styled.SecondColumn>
        <div>
          <span>Wind speed  <b>{(_.head(data)!.day.maxwind_kph / 3.6).toFixed(1)} m/s</b></span>
        </div>
        <Styled.FourthColumn>
          <span>{_.head(data)!.day.daily_chance_of_rain} %</span>
          <span>{_.head(data)!.day.daily_chance_of_snow} %</span>
        </Styled.FourthColumn>
      </Styled.WeatherToday>

      <Styled.WeatherWeek>
        {!loading && <>
          {
            _.map(data, (item, index) =>
              <div key={index}>
                <div>
                  {index === 0 ?
                    'Today'
                    : `${WEEK_S[new Date(item.date).getDay()]}, ${new Date(item.date).getDate()} `}
                </div>
                <Styled.WeatherCard>
                  <div>
                    <span>{item.day.maxtemp_c > 0 ? '+' + _.floor(item.day.maxtemp_c) : _.floor(item.day.maxtemp_c)}°</span>
                    <span>{item.day.mintemp_c > 0 ? '+' + _.floor(item.day.mintemp_c) : _.floor(item.day.mintemp_c)}°</span>
                  </div>
                  <img src={item.day.condition.icon} alt={item.day.condition.text} />
                </Styled.WeatherCard>
              </div>)
          }
        </>
        }

      </Styled.WeatherWeek>
    </Styled.WeatherData>
  )

}

export default WeatherBlock;
// export { }