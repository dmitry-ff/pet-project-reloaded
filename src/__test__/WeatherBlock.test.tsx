import WeatherBlock from '../Components/Weather/WeatherBlock';
import renderer from 'react-test-renderer';
import { TData } from '../Data/Types/TData';
import { shallow } from 'enzyme';
import { FirstColumn } from '../Style/WeatherStyle';
import _ from 'lodash';
const mockData = [
  {
    date: '2020-21-10',
    day: {
      maxtemp_c: 1,
      mintemp_c: 1,
      avgtemp_c: 1,
      maxwind_kph: 1,
      daily_chance_of_rain: 1,
      daily_chance_of_snow: 1,
      condition: {
        text: 'sunny',
        icon: 'icon.png',
      }
    }
  }
] as TData[];

describe('WeatherBlock component', () => {

  it('should render component with data array', () => {
    const tree = renderer
      .create(<WeatherBlock loading={false} nameHead={'Tula'} data={mockData} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render component', () => {
    const component = shallow(<WeatherBlock loading={false} nameHead={'Tula'} data={mockData} />)
    const wrapper = component.find('Styled(div)');
    expect(wrapper.length).toBe(7);
  })
})
