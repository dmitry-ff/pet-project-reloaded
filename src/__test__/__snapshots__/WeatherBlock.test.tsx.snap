// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherBlock component should render component with data array 1`] = `
<div
  className="css-foywk1"
>
  <span>
    Weather in 
    <b>
      Tula
    </b>
  </span>
  <div
    className="css-r06wbh"
  >
    <div
      className="css-1d01znx"
    >
      <span>
        +1
        °
      </span>
      <img
        alt="sunny"
        src="icon.png"
      />
    </div>
    <div
      className="css-1q8vl9a"
    >
      <span>
        sunny
      </span>
      <div>
        <span>
          max 
          <b>
            +1
          </b>
        </span>
        <span>
          min 
          <b>
            +1
          </b>
        </span>
      </div>
    </div>
    <div>
      <span>
        Wind speed  
        <b>
          0.3
           m/s
        </b>
      </span>
    </div>
    <div
      className="css-o65e2m"
    >
      <span>
        1
         %
      </span>
      <span>
        1
         %
      </span>
    </div>
  </div>
  <div
    className="css-1roh6vo"
  >
    <div>
      <div>
        Today
      </div>
      <div
        className="css-17p9zak"
      >
        <div>
          <span>
            +1
            °
          </span>
          <span>
            +1
            °
          </span>
        </div>
        <img
          alt="sunny"
          src="icon.png"
        />
      </div>
    </div>
  </div>
</div>
`;
